api/qanpb/profile.proto

```diff
@@ -39,32 +40,37 @@ message ReportMapFieldEntry {
 
 // ReportReply is list of reports per quieryids, hosts etc.
 message ReportReply {
-  repeated ProfileRow rows = 1;
+  uint32 total_rows = 1;
+  uint32 offset = 2;
+  uint32 limit = 3;
+  repeated Row rows = 4;
 }
 
-// ProfileRow define metrics for selected dimention.
-message ProfileRow {
+// Row define metrics for selected dimention.
+message Row {
   uint32 rank = 1;
-  float percentage = 2;
-  string dimension = 3;
-  float row_number = 4;
-  string d_servers = 5;
-  string d_databases = 6;
-  string d_schemas = 7;
-  string d_usernames = 8;
-  string d_client_hosts = 9;
-  string first_seen = 10;
-  float qps = 11;
-  float load = 12;
-  string fingerprint = 13;
-  Stats stats = 14;
+  string dimension = 2;
+  map<string, Metric> metrics = 3;
+  repeated Point sparkline = 4;
 }
 
-// Stats metrics.
-message Stats {
-  float num_queries = 1;
-  float m_query_time_sum = 2;
-  float m_query_time_min = 3;
-  float m_query_time_max = 4;
-  float m_query_time_p99 = 5;
+// Metric cell.
+message Metric {
+  Stat stats = 1;
+}
+
+// Stat is statistics of specific metric.
+message Stat {
+  float rate = 1;
+  float cnt = 2;
+  float sum = 3;
+  float min = 4;
+  float max = 5;
+  float p99 = 6;
+}
+
+// Point is a values of all selected metrics for one point of sparkline.
+// points: x -> timestamp or point, y -> values of any metric.
+message Point {
+  map<string, float> values = 1;
```

Comment:
Why not two separate float fields?

Reply:
it is not two separated fields. it will looks like:
```json
 "values": {
            "m_lock_time_sum": 0.00818,
            "m_query_load": 0.002203,
            "m_query_time_avg": 5.6115473e-06,
            "m_query_time_sum": 0.13218,
            "num_queries_sum": 23555,
            ...
            "point": 59,
            "time_frame": 60,
            "timestamp": 1546300800
          }
```

Reply:
Well, then comment is confusing. It probably should say "key" and "value" instead of "x" and "y".

Reply:
Improved comment. But not sure how this helps. Almost avery key can represent values for y coordinate.